### PR TITLE
fix(filesystem): fixed call to debounce with wrong arguments

### PR DESCRIPTION
**Pull request by pysan3**
closes: #1214

Please read [my comments](https://github.com/nvim-neo-tree/neo-tree.nvim/issues/1214#issuecomment-1807104855) for the explanation.

But ultimately, this is a failure of an untyped scripting language lol.

We definitely need to introduce type notations at some point. I'm too sick of these kind of errors.

### CODE REVIEW

**Comment by cseickel** [comment]
Actually, there are type annotations already:

https://github.com/nvim-neo-tree/neo-tree.nvim/blob/0f6e7acfd86b052acf78baccba04d5c61dcbbc0d/lua/neo-tree/utils/init.lua#L71-L76

The problem here is that both of the args are actually integers. We could add a validation on the value passed in for strategy and make sure that it does not overlap with valid values for frequency.

**Comment by pysan3** [comment]
There's a type specifically dedicated to this kind of problem. https://github.com/LuaLS/lua-language-server/wiki/Annotations#enum

I was kinda talking about type notation for the whole plugin. It is very easy to loose track of `state` as well as I hear a lot of questions of how to use `events` in the config.

We would want to define a `NeotreeState` [class](https://github.com/LuaLS/lua-language-server/wiki/Annotations#class) and also one for config would be nice.
Just like this: https://github.com/3rd/image.nvim/blob/master/lua/types.lua.

**Comment by cseickel** [comment]
> There's a type specifically dedicated to this kind of problem. https://github.com/LuaLS/lua-language-server/wiki/Annotations#enum

I didn't realize that was an option, can you apply that in this PR?

> Running a type check test for all PRs will solve most of the problems.
Like https://github.com/mrcjkb/lua-typecheck-action should do the job (haven't used myself tho).

Hell yeah! I'm adding it.


> I was kinda talking about type notation for the whole plugin. It is very easy to loose track of `state` as well as I hear a lot of questions of how to use `events` in the config.
> 
> We would want to define a `NeotreeState` [class](https://github.com/LuaLS/lua-language-server/wiki/Annotations#class) and also one for config would be nice. Just like this: https://github.com/3rd/image.nvim/blob/master/lua/types.lua.

That is an amazing idea, it would be great to have. I don't quite have that quantity of time any more but I would happily merge it if you (or anyone else) could do the work.

**Comment by pysan3** [comment]
I'll change it to the enum notation but I'd like to make a naming convention for the types we are going to be defining.

How about `NeotreePascalCase` for types that users might use outside (defined in `./neo-tree/types.lua` (Eg `NeotreeState`, `NeotreeConfig`, ...) and `neotree_snake_case` for types used inside? Eg, this enum will be named like `---@enum neotree_debounce_strategy`.

Unfortunately I'm quite busy now but I'll be able to work on it on February next year.

I don't think it's a good idea to add the action now since it will raise tons of errors all over the place.

If you are ok with waiting until next year, I can do the work then. It would be nice if we could do an entire feature freeze while I'm working on that since I think I'll need to change quite a lot of code.

On top of that, if that's ever possible I'm also thinking of reinventing a path module so that it works just right on all OS. As well as git on windows seems to be broken right now when there's a non-ascii char in the path.

**Comment by pysan3** [comment]
Here's a poc for enum. As you can see, diagnose is working correctly here ;)

image

**Comment by cseickel** [comment]
I think the Neotree prefix and using pascal case both sound good. I don't think it's worthwhile to have a different convention for internally used types though, I would just use Pascal case for all types.

You are right about the type checker. I opened a PR but there are a lot of errors and warnings to deal with so I don't know if I will be able to follow through or not.

I totally understand your time constraints, and if you are still willing and able to work on it next year then that will be great. I would not mind a feature freeze while you do that because I think that this work would be way more important than any new features.

regarding a path module, you're right about that. It might be a good idea to check on Plenary's Path module and see if it handles these issues or not.

**Comment by pysan3** [comment]
> I think the Neotree prefix and using pascal case both sound good. I don't think it's worthwhile to have a different convention for internally used types though, I would just use Pascal case for all types.

Alright, will go with that.

> You are right about the type checker. I opened a PR but there are a lot of errors and warnings to deal with so I don't know if I will be able to follow through or not.

I know how to fix the type issues so I can handle it if I've got time ;) (We have the final option `---@diagnostic disable-line` up our sleeves lol)

> I totally understand your time constraints, and if you are still willing and able to work on it next year then that will be great. I would not mind a feature freeze while you do that because I think that this work would be way more important than any new features.
> 

Thanks, I'll ping you when I can start working on it.

> regarding a path module, you're right about that. It might be a good idea to check on Plenary's Path module and see if it handles these issues or not.

Yah, I'm aware of that thing but I don't think it's meant for a heavy usage like in our case imo. It's lacking features and we would eventually have to implement many things with barebone luv anyways.

**Comment by pysan3** [comment]
@cseickel Do you know what you were trying to do here?
image

https://github.com/nvim-neo-tree/neo-tree.nvim/blob/0f6e7acfd86b052acf78baccba04d5c61dcbbc0d/lua/neo-tree/git/status.lua#L240

**Comment by pysan3** [comment]
> > I think the Neotree prefix and using pascal case both sound good. I don't think it's worthwhile to have a different convention for internally used types though, I would just use Pascal case for all types.
> 
> Alright, will go with that.

Well, we can distinguish between what would be considered as a _breaking change_ if we have some kind of difference between private types and public types.

Using private types from outside will be definitely valid but we can break it at any time, but for the pascal case ones, they should be announced as a breaking change or something like that.

**Comment by cseickel** [comment]
> @cseickel Do you know what you were trying to do here? image

That's just nonsense that slipped by me in #637. You can delete that line as its not doing anything.

**Comment by pysan3** [comment]
> That's just nonsense that slipped by me in #637. You can delete that line as its not doing anything.

Done.

I guess it's ready to go (tho the tests are failing...)

BTW may I ask your opinion on this? https://github.com/nvim-neo-tree/neo-tree.nvim/pull/1227#issuecomment-1807194997

**Comment by cseickel** [comment]
> Well, we can distinguish between what would be considered as a _breaking change_ if we have some kind of difference between private types and public types.
> 
> Using private types from outside will be definitely valid but we can break it at any time, but for the pascal case ones, they should be announced as a breaking change or something like that.

I just don't see types as something that are easily defined as internal or external. Methods and modules are what is truly internal/external. Besides, I'm not sure we actually have any types that would qualify as internal only:

https://github.com/nvim-neo-tree/neo-tree.nvim#what-is-a-breaking-change

File: lua/neo-tree/utils/init.lua [hunk 1]
```diff
@@ -70,9 +75,10 @@ end
 
 ---Call fn, but not more than once every x milliseconds.
 ---@param id string Identifier for the debounce group, such as the function name.
----@param fn function Function to be executed.
+---@param fn function? Function to be executed.
```
Comment (cseickel) [inline lua/neo-tree/utils/init.lua:78]: This is actually required.

Comment (pysan3) [reply lua/neo-tree/utils/init.lua]: Sorry, fixed.

**Comment by pysan3** [comment]
> > Well, we can distinguish between what would be considered as a _breaking change_ if we have some kind of difference between private types and public types.
> > Using private types from outside will be definitely valid but we can break it at any time, but for the pascal case ones, they should be announced as a breaking change or something like that.
> 
> I just don't see types as something that are easily defined as internal or external. Methods and modules are what is truly internal/external. Besides, I'm not sure we actually have any types that would qualify as internal only:
> 
> [nvim-neo-tree/neo-tree.nvim#what-is-a-breaking-change](https://github.com/nvim-neo-tree/neo-tree.nvim#what-is-a-breaking-change)

I see. I'll go all in with PascalCase :+1:

**Review by cseickel — APPROVED** [review]

**Comment by pysan3** [comment]
Not gonna start yet but pasting this as a reminder. 

https://github.com/mrcjkb/lua-typecheck-action/issues/34

**Pull request merged**